### DISCUSSION TITLE
Decoy v3

### DISCUSSION THREAD

**Discussion by mcous** [mcous/decoy#256] [Ideas]
Starting a discussion for the next major version of Decoy! Here's what's in my head so far

## Very likely

- No more rehearsal syntax
    - It complicates the architecture
    - It's primary intent was to preserve type-safety and intellisense, and `ParamSpec` solves this use case
    - Without rehearsals, Decoy is able to distinguish between calls and stub configurations, unblocking richer warnings and debugability
- Promote various warnings to errors as already called out in the docs
- Bump minimum supported version of Python

## Ideas / maybe

- `debug` helper like [vitest-when](https://github.com/mcous/vitest-when?tab=readme-ov-file#debugspy-tfunc-options-debugoptions-debugresult)
- Drop matchers in favor of [hamcrest](https://github.com/hamcrest/PyHamcrest) or other pre-existing library
- Drop global state
    - Might make call order verification difficult, and I think call order verification is valuable
- Drop custom mock object in favor of `MagicMock`?
    - Just a thought, not sure this is feasible or a good idea

**Comment by SyntaxColoring** [comment]
> * Promote various warnings to errors as already called out in the docs

Early heads-up that we have hundreds of `MissingSpecAttributeWarning`s that seem spurious. I haven't had a chance to look into them yet. If you have release timing in mind, I can try to investigate before then, in case it's a Decoy bug. (It might not be—it wouldn't surprise me if we're just speccing a bunch of mocks wrongly.)

> * Drop matchers in favor of [hamcrest](https://github.com/hamcrest/PyHamcrest) or other pre-existing library

This seems nice if the 3rd-party library is up to the task. Matchers seem like something that takes a surprising amount of care to handle all the edge cases.

Relatedly, I've been mixing `unittest.mock.sentinel` with Decoy and I haven't had any problems.

**Comment by mcous** [comment]
>> - Promote various warnings to errors as already called out in the docs
> Early heads-up that we have hundreds of MissingSpecAttributeWarnings that seem spurious.

😱. Let me know if you'd like any help looking into this. I peeped y'all's CI logs and, while my memory of the interfaces is not what it used to be, I wouldn't have expected to see those warnings

> If you have release timing in mind...

I have no release timing in mind whatsoever. Opentrons is likely the only real use of this library, so I'd only pursue working on it if it was helpful for y'all. Otherwise I'm not currently using Decoy, so I don't have any real pressure for a release.

> Relatedly, I've been mixing unittest.mock.sentinel with Decoy and I haven't had any problems.

Nice, yeah chalk that up as another point towards "what can Decoy re-use?"

**Comment by mcous** [comment]
Put a work in progress branch up! I'm happy with a lot of it, but the attribute mocking system needs another pass to remove its rehearsal system in favor of something simpler. It's also still missing some important features, like asynchronous context managers and matching function signatures.

https://github.com/mcous/decoy/compare/main...wip/decoy-3-api

It seems like I can probably release this as a preview (e.g. `from decoy.next import Decoy`) in v2 prior to cutting a v3 that drops the rehearsal syntax, old Python versions, etc.

**Comment by Alejandro-FA** [comment]
> Drop matchers in favor of [hamcrest](https://github.com/hamcrest/PyHamcrest) or other pre-existing library

Hey, just adding my two cents here. I agree that this might be beneficial in the long-run. However, it should be noted that hamcrest matcher methods return `Matcher[T]` objects, so it might not be possible to use them with stubs directly.

https://stackoverflow.com/questions/8348046/mockitos-matcher-vs-hamcrest-matcher

**Reply by mcous** [reply]
Oh damn, yeah I don't think it's possible (off the top of my head) to get that to play nicely with strict argument typing, even with the newer typing.Parameters

Might need to set the hamcrest idea aside. I keep running into frustrating roadblocks with Python's typing system on v3, so no sense trying to fit something in that won't fit cleanly.

I think there's some inspiration to be had both from your captor typing work and hamcrests' matched creation API to simplify our matchers though. Might also steal that mockito hamcrest wrapper mentioned in that SO post
